language/move-lang/src/hlir/translate.rs

```diff
@@ -0,0 +1,1227 @@
+
+                let mvar = |var, st| {
+                    let from_user = false;
+                    let mv = TE::Move { from_user, var };
+                    T::exp(N::Type_::single(st), sp(eloc, mv))
+                };
+                let econd = mvar(vcond, tu64);
+                let ecode = mvar(vcode, tbool);
+                let eabort = T::exp(tunit.clone(), sp(eloc, TE::Abort(Box::new(ecode))));
+                let eunit = T::exp(tunit.clone(), sp(eloc, TE::Unit));
+                let inlined_ = TE::IfElse(Box::new(econd), Box::new(eunit), Box::new(eabort));
+                let inlined = T::exp(tunit.clone(), sp(eloc, inlined_));
+                stmts.push_back(sp(eloc, T::SequenceItem_::Seq(Box::new(inlined))));
+
+                let block = T::exp(tunit, sp(eloc, TE::Block(stmts)));
+                return maybe_exp_(context, result, block);
+            }
+            let mut expected_types = single_types(context, parameter_types);
+            let expected_type = match expected_types.len() {
+                0 => sp(eloc, H::Type_::Unit),
+                1 => sp(eloc, H::Type_::Single(expected_types.pop().unwrap())),
+                _ => sp(eloc, H::Type_::Multiple(expected_types)),
+            };
+            let htys = base_types(context, type_arguments);
+            let harg = exp!(context, result, Some(&expected_type), *arguments);
+            let hacquires = acquires
+                .into_iter()
+                .map(|b| base_type(context, b))
+                .collect();
+            let call = H::ModuleCall {
+                module,
+                name,
+                type_arguments: htys,
+                arguments: harg,
+                acquires: hacquires,
+            };
+            HE::ModuleCall(Box::new(call))
+        }
+        TE::Builtin(bf, targ) => {
+            let arg = exp!(context, result, None, *targ);
+            builtin(context, result, eloc, *bf, arg)
+        }
+        TE::Dereference(te) => {
+            let e = exp!(context, result, None, *te);
+            HE::Dereference(e)
+        }
+        TE::UnaryExp(op, te) => {
+            let e = exp!(context, result, None, *te);
+            HE::UnaryExp(op, e)
+        }
+        TE::BinopExp(tl, op, tr) => {
+            let el = exp!(context, result, None, *tl);
+            let er = exp!(context, result, None, *tr);
+            HE::BinopExp(el, op, er)
+        }
+        TE::Pack(_, s, tbs, tfields) => {
+            let bs = base_types(context, tbs);
+
+            let decl_fields = context.fields(&s);
+            let decl_field = |f: &Field| -> usize { *decl_fields.get(f).unwrap() };
+
+            let mut texp_fields: Vec<(usize, Field, usize, N::BaseType, T::Exp)> = tfields
+                .into_iter()
+                .map(|(f, (exp_idx, (bt, tf)))| (decl_field(&f), f, exp_idx, bt, tf))
+                .collect();
+            texp_fields.sort_by(|(_, _, eidx1, _, _), (_, _, eidx2, _, _)| eidx1.cmp(eidx2));
```

Comment:
This idiom seems to be somewhat common and crop up in a number of places. Might be worthwhile pulling into a helper function.

Reply:
The sort by is always a different index, but I'll try to clean up the assign/unpack code